### PR TITLE
quic test timeout fix

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem

Still seeing timeout https://buildkite.com/solana-labs/solana/builds/86606#01850f20-ff56-4c79-bd59-3ca8f3f8a25b in quic streamer test. Locally it is hard to reproduce, it seems to show up in recent CI builds.

#### Summary of Changes

Allow longer chunk receive timeout without impacting testing the stream exit condition for unit tests. The exit is periodically checked, we will break only when the total allowed chunk receive timed out. The start time is reset when a new chunk is received.

Fixes #

### CODE REVIEW

File: streamer/src/nonblocking/quic.rs
Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs:570]: It should have no functional impact to the validator in runtime. The change still honors the timeout set by the input. It only wakes up multiple times to check if the exit condition is set so that we can quit in a more timely manner.

Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs:570]: The min is to guard against too small value which can wake up unnecessarily frequently and wasting CPU cycles. The max guard against waiting for too long which delay exit and cause some test failures when the timeout value is large. Within this value, the logic is to check 10 times for the set timeout if there are no data which I agree is heuristic.

Comment (pgarg66) [reply streamer/src/nonblocking/quic.rs:570]: Ok, that makes sense. Could you please add a comment in the code reflecting that? It may be hard to understand the reason for the future maintainers of the code.

File: streamer/src/nonblocking/quic.rs
Comment (ryleung-solana) [inline streamer/src/nonblocking/quic.rs:1030]: What problems are we seeing to cause us to want to raise this timeout? Is it just some CI tests? I'm concerned that this may lead to greater potential for denial of service...

File: streamer/src/nonblocking/quic.rs
Comment (ryleung-solana) [inline streamer/src/nonblocking/quic.rs:574]: Have we verified that if this times out, when we call stream.read_chunk again, it will try reading the same chunk? Have we tested this doesn't result in regressions in the case where we actually hit this timeout?

File: streamer/src/nonblocking/quic.rs
Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs:1030]: Yes -- we primarily see this in our CI builds. Hard to reproduce locally but been seen sporadically seen by folks including myself. The change itself should not increase any denial of service as it still honor the timeout requirement.

File: streamer/src/nonblocking/quic.rs
Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs:574]: This loop and on timeout trying to receive the chunk again is not new. The stream being splitter into multiple chunks and reassembled are old logic and it is already happening in production. See the code before https://github.com/solana-labs/solana/pull/28513. If times out reading a chunk, it will try to read a chunk again. There is not the concept of the "same" chunk though -- we did not receive a chunk and hence the timeout.

File: streamer/src/nonblocking/quic.rs
Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs:575]: Added

**Comment by pgarg66** [comment]
Looks good to me. Let's wait for @ryleung-solana to finish his review before merging it..

**Review by ryleung-solana — APPROVED** [review]

**Pull request merged**
